README.md
# Flightmon

A command-line "GUI" for watching ADS-B output from a dump1090 daemon.

![Screenshot of Flightmon in action](https://user-images.githubusercontent.com/390829/84444664-48db3080-ac10-11ea-994c-3910e9f382fd.png)

Test on Mac OS X, Linux amd64, and Linux arm (e.g. Raspberry Pi).

## Usage

To run it, just type `flightmon`. Your terminal will be taken over with a lovely table based on the SBS1 data stream coming from `localhost:30003`. You can specify a different host with `--dump_host`.

Full usage instructions:

```
Usage flightmon:
  -debug
        Enable debug output
  -dump_host string
        dump1090 SBS1 stream address (required) (default "localhost:30003")
  -max_age int
        Stop showing aircraft older than this many seconds (default 60)
  -show_ui
        Enable the UI (default true)
```

## Download

Head on over to the [releases page](https://github.com/mik3y/flightmon/releases) for pre-built binaries for Linux and Mac.

## Changelog

See [CHANGELOG.md](https://github.com/mik3y/flightmon/blob/master/CHANGELOG.md) for the latest changes.

## Contributing

Contributions welcome! Open an issue or a pull request if you've got ideas.


CHANGELOG.md
# Changelog

### v0.1.2 (2020-06-13)

* Internal cleanups.

### v0.1.1 (2020-06-12)

* Better error handling.

### v0.1.0 (2020-06-11)

* Initial release!


gui.go
package main

import (
	"fmt"
	"strings"
	"time"

	"github.com/gdamore/tcell"
	lru "github.com/hashicorp/golang-lru"
	pb "github.com/mik3y/goadsb/proto"
	"github.com/rivo/tview"
)

var updatedTextColor = tcell.ColorWhite
var normalTextColor = tcell.ColorGray

func getHeaderText(cache *lru.Cache) string {
	return fmt.Sprintf("Flightmon v%s · %d tracked", Version, cache.Len())
}

func updateRow(table *tview.Table, row int, positionData *pb.PositionUpdate) {
	var cell *tview.TableCell
	colNum := 0

	cell, colNum = table.GetCell(row, colNum), colNum+1
	cell.SetText(*positionData.IcaoId)

	now := int64(time.Now().UnixNano() / int64(time.Millisecond))
	ageSeconds := (now - *positionData.Timestamp) / 1000
	cell, colNum = table.GetCell(row, colNum), colNum+1
	cell.SetText(fmt.Sprint(ageSeconds))

	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Callsign != nil {
		if cell.Text != *positionData.Callsign {
			cell.SetText(*positionData.Callsign).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Lat != nil {
		s := fmt.Sprint(*positionData.Lat)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Lng != nil {
		s := fmt.Sprint(*positionData.Lng)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Altitude != nil {
		s := fmt.Sprint(*positionData.Altitude)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.GroundSpeed != nil {
		s := fmt.Sprint(*positionData.GroundSpeed)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Track != nil {
		s := fmt.Sprint(*positionData.Track)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.VerticalRate != nil {
		s := fmt.Sprint(*positionData.VerticalRate)
		if cell.Text != s {
			cell.SetText(s).SetTextColor(updatedTextColor)
		}
	}
	cell, colNum = table.GetCell(row, colNum), colNum+1
	if positionData.Squawk != nil {
		if cell.Text != *positionData.Squawk {
			cell.SetText(*positionData.Squawk).SetTextColor(updatedTextColor)
		}
	}
}

func clearUpdatedColors(table *tview.Table, row int) {
	for c := 0; c < table.GetColumnCount(); c++ {
		table.GetCell(row, c).SetTextColor(normalTextColor)
	}
}

func updateTable(cache *lru.Cache, table *tview.Table, header *tview.TextView) {
	// Clear any previously-updated cells.
	for r := 1; r < table.GetRowCount(); r++ {
		clearUpdatedColors(table, r)
	}

	// Pass 1: Refresh all existing rows.
	updatedMap := make(map[string]bool)
	for r := table.GetRowCount(); r >= 1; r-- {
		icaoID := table.GetCell(r, 0).Text
		updatedMap[icaoID] = true
		positionDataPtr, ok := cache.Peek(icaoID)
		if !ok {
			table.RemoveRow(r)
		} else {
			positionData := positionDataPtr.(*pb.PositionUpdate)
			updateRow(table, r, positionData)
		}
	}

	// Pass 2: Add any new rows.
	keyItems := cache.Keys()
	keys := make([]string, len(keyItems))
	for i, v := range keyItems {
		keys[i] = fmt.Sprint(v)
	}

	for idx := range keys {
		icaoID := keys[idx]
		if updatedMap[icaoID] {
			continue
		}
		positionDataPtr, ok := cache.Peek(icaoID)
		if !ok {
			continue
		}
		positionData := positionDataPtr.(*pb.PositionUpdate)
		rowNum := table.GetRowCount()
		for c := 0; c < table.GetColumnCount(); c++ {
			table.SetCell(rowNum, c, tview.NewTableCell("").SetTextColor(updatedTextColor))
		}
		updateRow(table, rowNum, positionData)
	}

	header.SetText(getHeaderText(cache))
}

// ShowUI shows a curses-like GUI in the terminal.
func ShowUI(cache *lru.Cache, donec chan<- struct{}) {
	app := tview.NewApplication()

	header := tview.NewTextView().
		SetText(getHeaderText(cache)).
		SetTextColor(tcell.ColorGreen)

	table := tview.NewTable().
		SetBorders(true)

	grid := tview.NewGrid().
		SetRows(1, 0).
		SetBorders(false).
		AddItem(header, 0, 0, 1, 1, 0, 0, false).
		AddItem(table, 1, 0, 1, 1, 0, 0, true)

	headers := strings.Split("ICAO Age Callsign Lat Lng Altitude GroundSpeed Track VerticalRate Squawk", " ")
	for h := 0; h < len(headers); h++ {
		table.SetCell(0, h,
			tview.NewTableCell(headers[h]).
				SetTextColor(tcell.ColorYellow).
				SetAlign(tview.AlignCenter))
	}

	table.SetFixed(1, 4)

	go func() {
		for {
			time.Sleep(time.Second)
			updateTable(cache, table, header)
			app.Draw()
		}
	}()

	if err := app.SetRoot(grid, true).SetFocus(table).Run(); err != nil {
		panic(err)
	}
	donec <- struct{}{}
}


sbs.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/golang/protobuf/proto"
	lru "github.com/hashicorp/golang-lru"
	"github.com/jpillora/backoff"
	pb "github.com/mik3y/flightmon/proto"
)

func safeParseInt(s string) (*int32, error) {
	if len(s) == 0 {
		return nil, nil
	}
	val, err := strconv.ParseInt(s, 10, 32)
	if err != nil {
		return nil, err
	}
	retVal := int32(val)
	return &retVal, nil
}

func safeParseFloat(s string) (*float32, error) {
	if len(s) == 0 {
		return nil, nil
	}
	val, err := strconv.ParseFloat(s, 32)
	if err != nil {
		return nil, err
	}
	retVal := float32(val)
	return &retVal, nil
}

func nilableString(s string) *string {
	if len(s) == 0 {
		return nil
	}
	trimmed := strings.TrimSpace(s)
	return &trimmed
}

func processMessage(message string, trackedCache *lru.Cache, updates chan<- *pb.PositionUpdate) error {
	msg, err := parseMessage(message)
	if err != nil {
		return err
	}

	changed := false

	ident := msg.GetHexIdent()
	record := &pb.PositionUpdate{
		IcaoId: &ident,
	}

	var existingRecord *pb.PositionUpdate

	val, exists := trackedCache.Peek(ident)
	if exists {
		existingRecord = val.(*pb.PositionUpdate)
	} else {
		logger.Infof("Tracking new aircraft: %v", ident)
		existingRecord = nil
		changed = true
	}

	if msg.Callsign != nil && (record.Callsign == nil || (*record.Callsign != *msg.Callsign)) {
		record.Callsign = msg.Callsign
		changed = true
	}

	if msg.Lat != nil && (record.Lat == nil || (*record.Lat != *msg.Lat)) {
		record.Lat = msg.Lat
		changed = true
	}

	if msg.Lng != nil && (record.Lng == nil || (*record.Lng != *msg.Lng)) {
		record.Lng = msg.Lng
		changed = true
	}

	if msg.Altitude != nil && (record.Altitude == nil || (*record.Altitude != *msg.Altitude)) {
		record.Altitude = msg.Altitude
		changed = true
	}

	if msg.GroundSpeed != nil && (record.GroundSpeed == nil || (*record.GroundSpeed != *msg.GroundSpeed)) {
		record.GroundSpeed = msg.GroundSpeed
		changed = true
	}

	if msg.Track != nil && (record.Track == nil || (*record.Track != *msg.Track)) {
		record.Track = msg.Track
		changed = true
	}

	if msg.VerticalRate != nil && (record.VerticalRate == nil || (*record.VerticalRate != *msg.VerticalRate)) {
		record.VerticalRate = msg.VerticalRate
		changed = true
	}

	if msg.Squawk != nil && (record.Squawk == nil || (*record.Squawk != *msg.Squawk)) {
		record.Squawk = msg.Squawk
		changed = true
	}

	now := int64(time.Now().UnixNano() / int64(time.Millisecond))
	record.Timestamp = &now

	if changed {
		updates <- record
	}

	if existingRecord != nil {
		proto.Merge(existingRecord, record)
		trackedCache.Add(ident, existingRecord)
	} else {
		trackedCache.Add(ident, record)
	}

	return nil
}

func parseMessage(message string) (*pb.SBS1Message, error) {
	parts := strings.Split(strings.TrimSuffix(message, "\n"), ",")

	if len(parts) != 22 {
		return nil, fmt.Errorf("Invalid message: %v", message)
	}

	var messageType pb.SBS1Message_MessageType
	switch parts[0] {
	case "SEL":
		messageType = pb.SBS1Message_SELECTION_CHANGE
		break
	case "ID":
		messageType = pb.SBS1Message_NEW_ID
		break
	case "AIR":
		messageType = pb.SBS1Message_NEW_AIRCRAFT
		break
	case "STA":
		messageType = pb.SBS1Message_STATUS_AIRCRAFT
		break
	case "CLK":
		messageType = pb.SBS1Message_CLICK
		break
	case "MSG":
		messageType = pb.SBS1Message_TRANSMISSION
		break
	}

	transmissionType, err := strconv.Atoi(parts[1])
	if err != nil {
		return nil, fmt.Errorf("Invalid transmissionType: %v", parts[1])
	}

	altitude, err := safeParseInt(parts[11])
	if err != nil {
		return nil, fmt.Errorf("Invalid altitude: %v", parts[11])
	}

	groundSpeed, err := safeParseInt(parts[12])
	if err != nil {
		return nil, fmt.Errorf("Invalid groundSpeed: %v", parts[12])
	}

	track, err := safeParseInt(parts[13])
	if err != nil {
		return nil, fmt.Errorf("Invalid track: %v", parts[13])
	}

	lat, err := safeParseFloat(parts[14])
	if err != nil {
		return nil, fmt.Errorf("Invalid lat: %v", parts[14])
	}

	lng, err := safeParseFloat(parts[15])
	if err != nil {
		return nil, fmt.Errorf("Invalid lng: %v", parts[15])
	}

	verticalRate, err := safeParseInt(parts[16])
	if err != nil {
		return nil, fmt.Errorf("Invalid verticalRate: %v", parts[16])
	}

	transmissionTypePtr := pb.SBS1Message_TransmissionType(transmissionType)

	p := pb.SBS1Message{
		MessageType:      &messageType,
		TransmissionType: &transmissionTypePtr,
		SessionId:        nilableString(parts[2]),
		AircraftId:       nilableString(parts[3]),
		HexIdent:         nilableString(parts[4]),
		FlightId:         nilableString(parts[5]),
		GeneratedDate:    nilableString(parts[6]),
		GeneratedTime:    nilableString(parts[7]),
		LoggedDate:       nilableString(parts[8]),
		LoggedTime:       nilableString(parts[9]),
		Callsign:         nilableString(parts[10]),
		Altitude:         altitude,
		GroundSpeed:      groundSpeed,
		Track:            track,
		Lat:              lat,
		Lng:              lng,
		VerticalRate:     verticalRate,
		Squawk:           nilableString(parts[17]),
	}

	return &p, nil
}

func startSbsTracking(trackedCache *lru.Cache, updates chan<- *pb.PositionUpdate) {
	if *dumpHost == "" {
		logger.Debug("No dump host specified, no updates will be posted")
		return
	}

	b := &backoff.Backoff{
		Min:    500 * time.Millisecond,
		Max:    10 * time.Second,
		Factor: 2,
		Jitter: false,
	}

	for {
		logger.Infof("Connecting to %s", *dumpHost)

		c, err := net.Dial("tcp", *dumpHost)
		if err != nil {
			logger.Errorf("Error connecting to %s: %v", *dumpHost, err)
			sleepAmount := b.Duration()
			logger.Errorf("Retrying in %s", sleepAmount)
			time.Sleep(sleepAmount)
			continue
		}
		defer c.Close()

		reader := bufio.NewReader(c)
		b.Reset()

		for {
			message, err := reader.ReadString('\n')
			if err != nil {
				logger.Errorf("Error reading from host: %s", err)
				break
			}
			err = processMessage(message, trackedCache, updates)
			if err != nil {
				logger.Errorf("Error processing message: %s", err)
				break
			}
		}
	}
}


main.go
package main

import (
	"flag"
	"os"
	"os/signal"
	"syscall"
	"time"

	lru "github.com/hashicorp/golang-lru"
	adsb "github.com/mik3y/goadsb"
	pb "github.com/mik3y/goadsb/proto"

	logging "github.com/ipfs/go-log"
)

var Version = "0.1.2"

var dumpHost = flag.String("dump_host", "localhost:30003", "dump1090 SBS1 stream address (required)")
var debug = flag.Bool("debug", false, "Enable debug output")
var showUI = flag.Bool("show_ui", true, "Enable the UI")
var maxAge = flag.Int("max_age", 60, "Stop showing aircraft older than this many seconds")

var logger = logging.Logger("main")

func getAgeOfUpdateInSeconds(update *pb.PositionUpdate) int {
	now := int64(time.Now().UnixNano() / int64(time.Millisecond))
	ageSeconds := (now - *update.Timestamp) / 1000
	return int(ageSeconds)
}

func startAging(cache *lru.Cache) {
	for {
		for _, key := range cache.Keys() {
			val, _ := cache.Peek(key)
			record := val.(*pb.PositionUpdate)
			age := getAgeOfUpdateInSeconds(record)
			if age >= *maxAge {
				logger.Debugf("Removing expired flight: %s", record.IcaoId)
				cache.Remove(key)
			}
		}
		time.Sleep(time.Second)
	}

}

func main() {
	flag.Parse()

	if *debug {
		logging.SetLogLevel("main", "debug")
	} else {
		if *showUI {
			logging.SetLogLevel("main", "warning")
		} else {
			logging.SetLogLevel("main", "info")
		}
	}

	trackedCache, err := lru.New(128)
	if err != nil {
		panic(err)
	}

	updates := make(chan *pb.PositionUpdate, 10)
	go adsb.StartSbsTracking(*dumpHost, logger.SugaredLogger, trackedCache, updates)
	go startAging(trackedCache)

	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT)

	donec := make(chan struct{}, 1)

	if *showUI {
		go ShowUI(trackedCache, donec)
	}

	for {
		select {
		case <-stop:
			os.Exit(0)
		case <-donec:
			os.Exit(0)
		case record := <-updates:
			if record == nil {
				os.Exit(0)
			}
			logger.Debugf("PositionUpdate: %v", record)
		}
	}
}
